### PR TITLE
Fixes #21388 - Filters for DockerManifestList

### PR DESCRIPTION
**Pull request by parthaa**
Adds code to set up docker content view filters that should work with
docker manifest lists. Also fixes a couple of bugs related to the docker
manifests filter generation code.

### CODE REVIEW

**Comment by bbuckingham** [comment]
@parthaa, While testing this PR, I had the observations below (which should be unrelated to these changes); however, wanted to get your input to see if we have or need additional issues to track changes in these areas.

* Content -> Content Views -> [view] -> Versions : shows a summary for Manifests, Manifest Lists and Tags; however, clicking on a Version does not.  Should it?
* Clicking on a version (e.g. version1.0) -> Docker Repositories : shows multiple repositories (one for archive and one for the env (Library)); however, the number of tags shown is inconsistent.  Is that expected?  If so, why?
  * Basically, the one associated with archive appears correct, but the environment does not
* Content -> Docker Tags : this page exists, but there is no similar page for Manifests or Manifest Lists.  Should there be?  Should there be any linkage of a Tag to a Manifest or List?
* Content -> Lifecycle Environments : selecting an environment, user can see Docker Tags, but nothing for Manifests or Lists.  Should they have access to them?

**Comment by bbuckingham** [comment]
@parthaa, Can you provide an example of how one can confirm that this is behaving as expected?

E.g. using busybox, we can specify any combination of the tags associated with it; however, it is unclear from the UI how the tag selected or result relates to a manifest list.

**Comment by parthaa** [comment]
Waiting on https://github.com/Katello/katello/pull/7066

**Comment by parthaa** [comment]
To Test this PR
- Sync busybox
- Goto the page listing the 'manifest lists' of the docker repo
- Make a note of the plausible tags (I choose glibc). Also note that the row containing glibc has 3 other tags.
- Click on the glibc tag and  a note that glic points to both a Manifest and a ManifestList
- Make copy of the ManifestList digest for glibc ```sha256:6cb02cbabacf59962d0cd05176e68ae4769e998e8d98f5e3f28fe915de341d6c```
- Now goto the page listing the 'manifests' of the docker repo. 
- Search for manifests belonging to that above list by entering ```manifest_list_digest =  sha256:6cb02cbabacf59962d0cd05176e68ae4769e998e8d98f5e3f28fe915de341d6c```
- Make a note of the 7 manifest items that come up in that search.
- Finally create a content view with the busy box repo
- Create a docker include filter - that says "Include both manifests and manifest lists pointed to by the tag glibc"
- Publish the content view

The resulting list should include

- 1 Docker Manifest List
- 4 Docker Tags (glibc + 3 others pointing to the same manifest list)
- 8 manifests -> 7 pointed to by the docker manifest list + 1 schema1 manifest pointed to by glibc

That pretty much is the new functionality. It copies the Manifest List based on the tag name and all the manifests pointed to by that list or tag name.

**Comment by parthaa** [comment]
> 
> Content -> Content Views -> [view] -> Versions : shows a summary for Manifests, Manifest Lists and Tags; however, clicking on a Version does not. Should it?
Thats a good point will be a better user experience to show the counts there. That being said I do not see this for yum repos either.  I think a new issue would make sense for this.
> Clicking on a version (e.g. version1.0) -> Docker Repositories : shows multiple repositories (one for archive and one for the env (Library)); however, the number of tags shown is inconsistent. Is that expected? If so, why?
Hmm I would not expect an inconsistency between the number of tags for library and archive version if library is supposed to have it. Looks like a bug.
> Basically, the one associated with archive appears correct, but the environment does not
Weird. That probably means promotion did not copy correctly. This PR should fix that aspect also. Can you resync and retest and confirm this behavior.
> Content -> Docker Tags : this page exists, but there is no similar page for Manifests or Manifest Lists. Should there be? Should there be any linkage of a Tag to a Manifest or List?
Manifests and lists by themselves are not as useful as the Tags. That being said it might be useful to show them. RFE's for these might be a good idea.
> Content -> Lifecycle Environments : selecting an environment, user can see Docker Tags, but nothing for Manifests or Lists. Should they have access to them?
Tags are probably the most useful things to search for w.r.t to docker. I don't see much use with manifests and manifest list searching there. If/When pulp implements docker mirror on sync we'll probably not have any manifests that do not have a tag. It good to have it in a Version Details screen but not top level imo.

@thomasmckay @jlsherrill feel free to add your opinions here.

**Comment by bbuckingham** [comment]
> > 
> > Content -> Content Views -> [view] -> Versions : shows a summary for Manifests, Manifest Lists and Tags; however, clicking on a Version does not. Should it?

> Thats a good point will be a better user experience to show the counts there. That being said I do not see this for yum repos either.  I think a new issue would make sense for this.

A new issue is totally ok.

> > Clicking on a version (e.g. version1.0) -> Docker Repositories : shows multiple repositories (one for archive and one for the env (Library)); however, the number of tags shown is inconsistent. Is that expected? If so, why?

> Hmm I would not expect an inconsistency between the number of tags for library and archive version if library is supposed to have it. Looks like a bug.

> > Basically, the one associated with archive appears correct, but the environment does not

> Weird. That probably means promotion did not copy correctly. This PR should fix that aspect also. Can you resync and retest and confirm this behavior.

It looks like this behavior may still exist.  For the version that I currently have published in Library (version 5.0) looking at 'version 5.0' -> 'Docker Repositories' shows 2 repos:

   default_organization-docker_busybox-5.0-docker-busybox	4 (tags)
   default_organization-library-docker_busybox-docker-busybox	75 (tags)

For a version that was promoted to 'test' (version 3.0) looking at 'version 3.0' -> 'Docker Repositories' shows 2 repos with the expected number of tags:

   default_organization-test-docker_busybox-docker-busybox	4 (tags)
   default_organization-docker_busybox-3.0-docker-busybox	4 (tags)

Is the above correct and expected?

> > Content -> Docker Tags : this page exists, but there is no similar page for Manifests or Manifest Lists. 

> Should there be? Should there be any linkage of a Tag to a Manifest or List?
Manifests and lists by themselves are not as useful as the Tags. That being said it might be useful to show them. RFE's for these might be a good idea.

> > Content -> Lifecycle Environments : selecting an environment, user can see Docker Tags, but nothing for Manifests or Lists. Should they have access to them?

> Tags are probably the most useful things to search for w.r.t to docker. I don't see much use with manifests and manifest list searching there. If/When pulp implements docker mirror on sync we'll probably not have any manifests that do not have a tag. It good to have it in a Version Details screen but not top level imo.

**Comment by parthaa** [comment]
> For a version that was promoted to 'test' (version 3.0) looking at 'version 3.0' -> 'Docker Repositories' shows 2 repos with the expected number of tags:
> 
> default_organization-test-docker_busybox-docker-busybox	4 (tags)
> default_organization-docker_busybox-3.0-docker-busybox	4 (tags)
> 
> Is the above correct and expected?

I created a separate issue for this bug because it affects downstream and we will want to track/address it separately -> http://projects.theforeman.org/issues/21808

**Comment by parthaa** [comment]
> Per the discussion, I would recommend creating an issue to show consistent summary on:
> Content -> Content Views -> [view] -> Versions and then when clicking on a Version (e.g. version 1.0)

Created http://projects.theforeman.org/issues/21828 to address this.
